core/src/main/java/io/grpc/transport/AbstractClientStream.java

```diff
@@ -173,6 +173,7 @@ protected void inboundTrailersReceived(Metadata.Trailers trailers, Status status
     this.status = status;
     this.trailers = trailers;
     deframe(ReadableBuffers.empty(), true);
+    inboundPhase(Phase.STATUS);
```

Comment:
This should be set as early as possible (so immediately after the if). Otherwise any reentrant calls will see the wrong phase. That is already being for the other inboundPhase()s in this class.